### ISSUE TITLE
Getting different output from the PyTorch model using the same input

### ISSUE THREAD

**Issue by shoaib6174** [shoaib6174/GSOC-22-Video-Swin-Transformers#5]
@sayakpaul @gustheman 

The PyTorch model generates different outputs with the same input. 

I have compared the outputs of two attempts (attempt 0 and attempt 2) with the same input using numpy.testing.assert_allclose and got the following-

![image](https://user-images.githubusercontent.com/40586752/185849454-3299e9bd-07c8-4885-a20c-82015f819a39.png)

But there are times when out of 10 attempts I got the same Final Output for multiple attempts. Here attempt 0, 3, 7, and 8 has the same Final Output.  

![image](https://user-images.githubusercontent.com/40586752/185848138-5add8af6-c8ce-4f7f-9aa2-2c76d9773a7a.png)

To understand the issue, I have inspected the layer-wise outputs. The output of the PatchEmbed3D layer, Basic Layer 1, and Basic Layer 2 produces similar output at each attempt. Though the second basic layer sometimes gives different output.

![image](https://user-images.githubusercontent.com/40586752/185845741-00047c7e-abac-44dc-8bfc-b9dc5257f9d8.png)

[ Here I have passed the same input through the PyTorch model 4 times, extracted the output of each layer at every attempt, and then compared them]

The deviation starts occurring after the Basic layer2. The basic layer3 provides different outputs most of the time. Here for attempts 2, 3 Basic Layer3 has provided the same output but different output for attempts 0 and 1.  The following layers also do the same. 
![image](https://user-images.githubusercontent.com/40586752/185848102-98989327-e54e-4105-b9b2-95f73e98c28f.png)

 

### **About the TensorFlow model**, 

The output of the PatchEmbed3D layer of both of the models matches. 

![image](https://user-images.githubusercontent.com/40586752/185848166-122dfde9-6bef-4527-99e9-9bf115807110.png)

[Here I have passed the same input through both of the models after assigning the PyTorch model's weights to the TF model and extracted the output of each layer. Then I compared them]

In the Basic layer 1 &  2 there is a slight deviation in the output values-

![image](https://user-images.githubusercontent.com/40586752/185848203-2a9294fd-4126-4b5f-a0ac-e80649b930cc.png)

This variation increases in the subsequent layers-

![image](https://user-images.githubusercontent.com/40586752/185848272-e56880e5-4956-45b4-a077-9e41ba660089.png)
 
And when it comes to the final output they seem totally different.

 But sometimes I have got similar Final Output from them-

![image](https://user-images.githubusercontent.com/40586752/185848311-4b0a1970-1ffa-434c-895e-e3ccec98dc28.png)
Here you can see the final output of the two models has only slight variation.

But if I run these two models again with the same input as this, I will get a different output for the PyTorch model. 

![image](https://user-images.githubusercontent.com/40586752/185848334-459f19c5-1421-46f3-baf5-2f02df73b4f7.png)

NB: The TF model's outputs are the same for each attempt with the same input

I think the reason behind this could be floating point calculation.

Here is the colab's link for your reference -

https://colab.research.google.com/drive/1N6ZJKBz_Ha4MYi05Sv_OgujL3GHQoKk5?usp=sharing

**Please have a look and suggest to me what I should do about it.**

**Comment by sayakpaul** [comment]
@shoaib6174 some thoughts:

* Could it be because of the wrong transposition format for the Conv3D params?
* I would start with the first layer from the PT and TF model and see if the outputs match. If they don't, we have a problem. You could extend this to the individual layers and see if that's a problem. Of course, the layers need to be having the same params ported inside of them, otherwise, the comparison will be meaningless. 
* Please ensure you're following the exact same defaults the PT model is using for the TF model. 
* Also, `get_x()` returns outputs that have `channels_first` memory layout. For TF models, we would want to have `channels_last` memory layout. This impacts how the layer params are initialized inside the TF model.

**Comment by shoaib6174** [comment]
@sayakpaul 

Here I have printed the layerwise output of both models-
![image](https://user-images.githubusercontent.com/40586752/185852220-e2f7fcbd-881a-4e89-9516-0b82cef1f1a8.png)

The output of the patch_embed layer matches exactly. So the Conv3D layer is not the problem.

About the input format- I have transposed the input before passing through the Conv3D layer. Here is the code-
![image](https://user-images.githubusercontent.com/40586752/185853134-023bc2ef-4372-4bf7-bb1b-3955eec87a57.png)


The basic layer1 of both models gets the same input from the patch_embed layer and creates different outputs.  If you look at the output values of basic layer1- the variation in the values is very small. 

I get this comparing the output of "basic layer1" of both models
![image](https://user-images.githubusercontent.com/40586752/185852618-54631906-42e0-4ee8-8b79-920c94e865c5.png)

**Comment by shoaib6174** [comment]
@sayakpaul 

Our main issue is that the PyTorch model gives different outputs for the same input. 

Here you can see that I am passing the same input through the PyTorch Model for 10 attempts and storing the outputs of each layer in outputs.
![image](https://user-images.githubusercontent.com/40586752/185854006-08d3e3ca-9eab-470b-9717-884f07d44d23.png)

All attempts give the same result for the patch_embed layer.
![image](https://user-images.githubusercontent.com/40586752/185854134-da74da32-1cd9-4039-91fe-da9530fc7f38.png)

The outputs of all attempts are also the same for "basic layer1"
![image](https://user-images.githubusercontent.com/40586752/185854219-d9835344-6ae7-4d4a-a36f-7a3c2f38eb40.png)

For "basic layer2" the output of attempt 5 is differnt from others though it received the same input as others- 
![image](https://user-images.githubusercontent.com/40586752/185854427-8f8570ea-8eb8-479e-a6db-eaf40b8a5c7d.png)

The situation gets worse at "basic layer3". The output of attempts 5,6,8,9 are different- 
![image](https://user-images.githubusercontent.com/40586752/185854607-de3d427f-83ea-4d27-8965-dd00422ba9dd.png)


And when we go to the final_output the output of only attempt 0,1,3,4 are the same. The rest is different. 
![image](https://user-images.githubusercontent.com/40586752/185854878-dd444b23-966e-4988-a82a-1b97f5c6e12b.png)

So we get different final outputs for the same input in the PyTorch model. And we can't say it's because of any random number initialization as the final outputs of four attempts have matched.

What could be the reason behind this?

**Comment by shoaib6174** [comment]
@sayakpaul 

In the case of TensorFlow, the same input always creates the same output. The output of the PyTorch model is inconsistent.

**Comment by sayakpaul** [comment]
What happens if the random generator seed is fixed for PyTorch?

**Comment by shoaib6174** [comment]
@sayakpaul  I have solved the issue of PyTorch Model creating different output with same input values. I had to use model.eval() after loading the model. Now each attempt producing same result. 

This result has some some difference with the TF output. `np.testing.assert_allclose()` with atol and rtol set to **1e-1** asserts the following- 

<img width="928" alt="image" src="https://user-images.githubusercontent.com/40586752/185898617-7ab6e9ae-0867-4833-ae1c-6cf255598324.png">

If I set atol and rtol to **1e-4** then it asserts the following-
<img width="928" alt="image" src="https://user-images.githubusercontent.com/40586752/185899481-31fc4304-ae90-43d1-ba59-b124adadc294.png">

And here is the layerwise output of both model-
<img width="767" alt="image" src="https://user-images.githubusercontent.com/40586752/185899660-4dcbf9f5-0644-45fe-a471-0c26b58d8c19.png">

**Comment by sayakpaul** [comment]
We can' settle with `1e-1`, it's too high of a tolerance here. 

Also, when comparing outputs, set the gradient computation for PT to False `torch.set_grad_enabled(False)`. With this, you won't have to do `with torch.no_grad()`.

**Comment by shoaib6174** [comment]
@sayakpaul 
Now the situation is basic layer1 of both models gets the same input and it produces different output. For testing purpose I'm using **1e-2** .
<img width="1292" alt="image" src="https://user-images.githubusercontent.com/40586752/185902919-35204647-547b-439e-8d11-af657ceffdb1.png">

Here 33% elements are mismatched. In the subsequent layers it creases and reaches 72% for the final layer.

Could the floating point calculation be the reason behind this?

**Comment by sayakpaul** [comment]
Please keep the tolerance to a minimum of 1e-4. 

> Could the floating point calculation be the reason behind this?

Not to this extent. Now that we know where the mismatch stems from (BasicLayer), I guess you can look into the implementation and correct what's missing?

**Comment by shoaib6174** [comment]
Okay, Dada. I will look into the basic layer. Thank you so much.
